### PR TITLE
Jacobian prototype (sparsity)

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes part of https://github.com/Deltares/Ribasim/issues/419.

### CODE REVIEW

**Comment by visr** [comment]
Great! Did you do any benchmarks, like https://github.com/Deltares/Ribasim/issues/419#issuecomment-1649824090? Would be interesting to compare. Why did you go for using it only when `density <= 0.001`, is slower for non-dense otherwise? I expect real models to generally meet that criterion, but testmodels not.

**Comment by visr** [comment]
Would be great if you could still do those benchmarks to know how much this helps #419.

For simplicity it's also fine to always be sparse, since only models that benefit from it are problematically slow. And potentially it could be harder to debug if a change to the model causes a different solver method to be used.

**Comment by SouthEndMusic** [comment]
Benchmarks: https://github.com/Deltares/Ribasim/issues/419#issuecomment-1651412616

**Comment by evetion** [comment]
Is there some way we can assert or test that the sparse matrix is correct? Does DiffEQ error if we set wrong values?

Can we maybe do a count (sum), and compare that with the number of (specific) edges?

I'm uncertain about adding to/changing this function, especially if any mistakes could lead to silent corruption.

**Comment by evetion** [comment]
Maybe we can rewrite by adding a function to parameters that enumerates the connections for each specific nodetype. Adding a new nodetype would then only require adding a method that determines whether that node is needed in this matrix, instead of editing the jac_prototype (which can be very small and generic).

**Comment by SouthEndMusic** [comment]
> Is there some way we can assert or test that the sparse matrix is correct? Does DiffEQ error if we set wrong values?

I do not directly see how. You could do a finite difference approximation of the Jacobian and see where the non-zeros are, but that is doomed to fail because of terms that are only sometimes constant, changes in which nodes are active, etc.

All tests now use the `jac_prototype`. When I was still debugging it I only got error messages when tests failed, the errors in the `jac_prototype` were not explicitly detected anywhere.

> Can we maybe do a count (sum), and compare that with the number of (specific) edges?

Maybe, but the way `jac_prototype` specifically works by finding edge types (or small subgraphs when `FractionalFlow` is involved). Only basins that have a direct dependency between them affect the Jacobian, not for instance edges connected to boundary nodes.

**Comment by SouthEndMusic** [comment]
@huite @evetion I refactored the code and added quite a bit of documentation, let me know whether it is clear now.

@evetion there is a problem with counting the number of non-zeros in `jac_prototype` and comparing that to how many times you find dependencies between basins, and it is with situations like this:

rating_curve_model

e.g. basins can depend on each other in more than one way.

File: core/src/utils.jl [hunk 1]
```diff
@@ -391,3 +391,191 @@ function expand_logic_mapping(
...
+    # One-directional flow describing nodes; contributions to the Jacobian
+    # if the unique nodes upstream and possibly multiple noes downstream are basins
+    update_jac_prototype!(jac_prototype, p, pump)
+    update_jac_prototype!(jac_prototype, p, tabulated_rating_curve)
+
+    # PidControl nodes; the controlled basin affects itself
+    # and the basins upstream and downstream of the controlled node
+    # affect eachother (if there is a basin upstream of the pump)
+    # For explanation for the integral term see below
+    update_jac_prototype!(jac_prototype, p, pid_control)
```
Comment (evetion) [inline core/src/utils.jl:428]: We could further generalize by something like

```julia
nodefields(p::Parameters) = (
    name for
    name in fieldnames(typeof(p)) if fieldtype(typeof(p), name) <: AbstractParameterNode
)

for nodefield in nodefields(p)
    update_jac_prototype!(jac_prototype, p, getfield(p, nodefield)
end
```

File: core/src/utils.jl [hunk 2]
```diff
@@ -391,3 +391,191 @@ function expand_logic_mapping(
...
+
+"""
+Set a value in the Jacobian prototype after checking it wasn't already set
+"""
+function set_nonzero!(
+    jac_prototype::SparseMatrixCSC{Float64, Int64},
+    i::Int,
+    j::Int,
+)::Nothing
+    msg = "Attempt at setting Jacobian nonzero twice for indices $i,$j."
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply core/src/utils.jl:441]: I think `LazyString` is better

**Review by evetion — APPROVED** [review]

**Pull request merged**